### PR TITLE
Add tests for some grading/ functions, and better docs for white-diff

### PR DESCRIPTION
**Pull request by stefano-maggiolo**
This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/830)

### CODE REVIEW

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for common grading functions.
+
+"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Add:
```
from __future__ import division
```
and
```
from future.builtins.disabled import *
from future.builtins import *
```

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import StringIO
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Use the one from the `io` module.

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        self.assertEqual(format_status_text(["你好"]), "你好")
+
+    def test_success_with_placeholders(self):
+        self.assertEqual(format_status_text(["%s", "ASD"]), "ASD")
+        self.assertEqual(format_status_text(["ASD%s\n%s", "QWE", "123"]),
+                         "ASDQWE\n123")
+
+    def test_success_with_translator(self):
+        # Translation is applied before formatting.
+        self.assertEqual(format_status_text(["A%s", "ASD"], self._tr), "EASD")
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Maybe also translator w/o placeholder?

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        self.assertEqual(
+            format_status_text(["AAA %s\n%s", "AAA", "AE"], self._tr),
+            "EEE AAA\nAE")
+
+    def test_insuccess(self):
+        self.assertEqual(format_status_text(["%s"]), "N/A")
+        self.assertEqual(format_status_text(["%s"], self._tr), "N/E")
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Maybe empty list too?

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 1 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done.

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            "exit_status": exit_status,
+        }
+        if signal is not None:
+            r["signal"] = signal
+        if filename is not None:
+            r["filename"] = filename
+        return r
+
+    def test_success_status_ok(self):
+        self.assertEqual(
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Isn't it bad practice to compare floats for exact equality? Looking at the doc I see that starting with Python 3.1 special type-specific comparators are used. So maybe here we use a dict comparator, that then in turn uses a float approximate comparator? Could you check?

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd"),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd"))
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="9"),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="9"))
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Maybe add one where the second status is ok?

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            r0, self._res(1.0, 2.0, 300, Sandbox.EXIT_OK))
+        self.assertEqual(
+            r1, self._res(0.1, 0.2, 0.3, Sandbox.EXIT_SIGNAL, signal="11"))
+
+
+class TestWhiteDiff(unittest.TestCase):
+
+    @staticmethod
+    def _diff(s1, s2):
+        return white_diff(StringIO.StringIO(s1), StringIO.StringIO(s2))
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: White diff is designed to operate on bytes, not strings.

Comment (lerks) [reply cmstestsuite/unit_tests/grading/__init__.py]: Which also means that all string literals in this test should become `b"..."`. (I guess right now an implicit encoding using utf-8 is happening somewhere)

Comment (lerks) [reply cmstestsuite/unit_tests/grading/__init__.py]: About that, maybe add some tests that use non-ascii-encoded strings...

File: docs/Task types.rst [hunk 8]
```diff
@@ -92,3 +92,20 @@ The resulting executable is run twice (one acting as the computer, one acting as
...
 
 Normally, the standard output of the second invocation of the manager is compared to a provided reference output file using the white-diff comparator. However, the admins may provide a :file:`checker` executable, with the same properties as for Batch. If a file with such a name is found in the uploaded manager files, it will be run instead of the white-diff comparator.
+
+
+.. _tasktypes_white_diff:
+
+White-diff comparator
+=====================
+
+White-diff is the only built-in comparator, and is used in most task types. It can be used when each testcase has a single correct output files, up to whitespaces.
```
Comment (lerks) [inline docs/Task types.rst]: s/single/unique/?

and s/files/file/

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 3 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 4 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done.

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 5 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done. The dict-specific is the same as assertEqual but in addition tests they are dicts, so no special float comparison.

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 6 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done.

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 7 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done.

File: docs/Task types.rst (hunk 8 above)
Comment (stefano-maggiolo) [reply docs/Task types.rst]: Done

**Comment by stefano-maggiolo** [comment]
PTAL

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 9]
```diff
@@ -22,10 +22,16 @@
...
 from __future__ import print_function
 from __future__ import unicode_literals
+from future.builtins.disabled import *
+from future.builtins import *
 
-import StringIO
+try:
+    from BytesIO import BytesIO
+except ImportError:
+    from io import BytesIO
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: No need: `io.BytesIO` exists in both py2 and py3.

File: cmstestsuite/unit_tests/grading/__init__.py [hunk 10]
```diff
@@ -77,74 +90,94 @@ def _res(execution_time, execution_wall_clock_time, execution_memory,
             r["filename"] = filename
         return r
 
+    def assertRes(self, r0, r1):
+        """Assert that r0 and r1 are the same result."""
+        self.assertAlmostEqual(r0["execution_time"], r1["execution_time"])
```
Comment (lerks) [inline cmstestsuite/unit_tests/grading/__init__.py]: Would you mind adding something equivalent to `assert {"execution_time", "execution_wall_clock_time", "execution_memory", "syscall", "signal", "filename"}.issubset(iterkeys(r0))` and similarly for `r1`?

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 9 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Done.

File: cmstestsuite/unit_tests/grading/__init__.py (hunk 10 above)
Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Apart that it'd be wrong (the last three are usually not present), I disagree. I might add set(iterkeys(r0)) == set(iterkeys(r1)), if that's your goal?

Comment (stefano-maggiolo) [reply cmstestsuite/unit_tests/grading/__init__.py]: Agreed offline not to add anything.

**Pull request merged**